[KCC] Solution demonstrating multi-cluster ingress using NEGs and auto-NEG controller

AlexBulankou opened the pull request:
This solution uses [Autoneg controller](https://github.com/GoogleCloudPlatform/gke-autoneg-controller) and Helm for customization.
FYI @soellman

maqiuyujoyce commented on config-connector/solutions/networking/helm/mci/README.md:
````diff
@@ -0,0 +1,153 @@
+1. Cloud Resource Manager API needs to be enabled on the project to use [ServiceUsage Resource](https://cloud.google.com/config-connector/docs/reference/resources#service). You can enable it by running:
+
+    ```bash
+    gcloud services enable cloudresourcemanager.googleapis.com --project [PROJECT_ID]
````
It's not clear which project to enable this api: The project where Config Connector is installed or the project where Config Connector is going to manage resources in?

morgante replied:
+1, also could we simply include this in the declarative config?

AlexBulankou replied:
This is actually not needed, deleting. But to respond to @morgante's comment, this is a special one that is [needed](https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/135).

maqiuyujoyce commented on config-connector/solutions/networking/helm/mci/README.md:
````diff
@@ -0,0 +1,153 @@
+1. Subsequent steps will use $PROJECT_ID variable
+
+    ```bash
+    export PROJECT_ID=[PROJECT_ID]
````
It's not clear which project to set here: The project where Config Connector is installed or the project where Config Connector is going to manage resources in?

AlexBulankou replied:
Added note.

maqiuyujoyce commented on config-connector/solutions/networking/helm/mci/README.md:
```diff
@@ -0,0 +1,153 @@
+
+All steps are run from this directory.
+
+1. Review and update the values in `./values.yaml`.
```
There is no `./values.yaml` file. Which values file should this be?

BrianPi replied:
It might be good to add a note that changing cluster name values in `./lb/values.yaml` will change the corresponding names needed below.

AlexBulankou replied:
Fixed. And added the note.

BrianPi commented on config-connector/solutions/networking/helm/mci/README.md:
```diff
@@ -0,0 +1,153 @@
+
+1. Review and update the values in `./values.yaml`.
+
+1. Subsequent steps will use $PROJECT_ID variable
```
I think requiring user-set environment variables is a discouraged pattern.

AlexBulankou replied:
Fixed

BrianPi commented on config-connector/solutions/networking/helm/mci/README.md:
````diff
@@ -0,0 +1,153 @@
+    export PROJECT_ID=[PROJECT_ID]
+    ```
+
+1. Configure clusters and load balancing resources with Helm:
````
To match with the other Helm solutions, you can remove the `$` in the bash below, and add something like "where `PROJECT_ID` is the project you want to deploy the load balancer in."

AlexBulankou replied:
Fixed, no longer using $PROJECT_ID

BrianPi commented on config-connector/solutions/networking/helm/mci/README.md:
```diff
@@ -0,0 +1,153 @@
+    gcloud container clusters get-credentials [CLUSTER NAME] --zone=[CLUSTER ZONE]
+
+    # curl the external address of the forwarding rule. Note that it might take around 5-10 minutes for load balancing to start working.
+    # You will see the message ("Hello from North America" or "Hello from Europe" backed on your location).
```
"based on your location"?

AlexBulankou replied:
Good catch! fixed.

BrianPi commented on config-connector/solutions/networking/helm/mci/lb/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+description: A Helm chart to deploy load balancing resources for GCP multi-cluster ingress
+name: gcp-mci-lb
+version: 0.1.0
+icon: https://upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Globe_icon.svg/1024px-Globe_icon.svg.png
```
Other solutions have abstained from providing any icon... should we use this one generally?

AlexBulankou replied:
There's also an icon in wordpress solution. I would leave this one to the author, not try to standardize.

morgante commented on config-connector/solutions/networking/helm/mci/README.md:
```diff
@@ -0,0 +1,153 @@
+# Multi-cluster ingress
```
We should call this "autoneg" or "standalone neg."

Multicluster ingress is already used to refer to the [native solution](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress-for-anthos).

AlexBulankou replied:
Changed

morgante commented on config-connector/solutions/networking/helm/mci/README.md:
```diff
@@ -0,0 +1,153 @@
+
+Multi-cluster ingress solution uses network endpoint groups to provision load balancing across multiple clusters.
+This solution uses [gke-autoneg-controller](https://github.com/GoogleCloudPlatform/gke-autoneg-controller).
+For demonstration purposes it uses a docker image with a simple Node.js service (bulankou/node-hello-world) with a single endpoint that prints out a message.
```
Any reason to use a custom image instead of `nginx`? (The latter is more likely to be available in clusters or easily retrieved).

AlexBulankou replied:
I would like an easy way to return a message on response, the message that I can specify in ENV parameters to the pod, not very familiar with nginx, how can I do it easily:
```
- name: node-app-container
        image: bulankou/node-hello-world
        env:
        - name: HELLO_MESSAGE
          value: {{ .Values.localMessage}}
        ports:
        - containerPort: 8080
```

morgante replied:
Ah, I missed that the first time around.

You could do this with nginx + configmap, but I agree it's probably not worth changing now.

morgante commented on config-connector/solutions/networking/helm/mci/cluster/templates/autoneg.yaml:
```diff
@@ -0,0 +1,209 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
I assume this is pulled from https://github.com/GoogleCloudPlatform/gke-autoneg-controller/tree/master/deploy (which we should convert into a kpt package).

AlexBulankou replied:
Made consistent. And added clarification in readme, to say which file we are using from https://github.com/GoogleCloudPlatform/gke-autoneg-controller.

morgante commented on config-connector/solutions/networking/helm/mci/lb/templates/gcp-cluster-1.yaml:
```diff
@@ -0,0 +1,39 @@
+          name: cluster-1-secret-pwd
+          key: password
+    clientCertificateConfig:
+      issueClientCertificate: false
```
+1,  I think it's a good idea to have a newline at the end of all files.

AlexBulankou replied:
Fixed

AlexBulankou commented on config-connector/solutions/networking/helm/mci/README.md:
````diff
@@ -0,0 +1,153 @@
+    export PROJECT_ID=[PROJECT_ID]
+    ```
+
+1. Configure clusters and load balancing resources with Helm:
````
Fixed.

AlexBulankou commented on config-connector/solutions/networking/helm/mci/cluster/templates/autoneg.yaml:
```diff
@@ -0,0 +1,209 @@
+          requests:
+            cpu: 100m
+            memory: 20Mi
+      terminationGracePeriodSeconds: 10
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/mci/cluster/templates/deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/mci/cluster/templates/service.yaml:
```diff
@@ -0,0 +1,29 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/mci/lb/templates/gcp-cluster-2.yaml:
```diff
@@ -0,0 +1,39 @@
+          name: cluster-2-secret-pwd
+          key: password
+    clientCertificateConfig:
+      issueClientCertificate: false
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/mci/lb/templates/gcp-custom-role-autoneg.yaml:
```diff
@@ -0,0 +1,25 @@
+    - compute.backendServices.get
+    - compute.backendServices.update
+    - compute.networkEndpointGroups.use
+    - compute.healthChecks.useReadOnly
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/mci/lb/templates/gcp-service-account-autoneg.yaml:
```diff
@@ -0,0 +1,20 @@
+metadata:
+  name: autoneg-system
+spec:
+  displayName: Autoneg Service Account
```
fixed

AlexBulankou commented:
> Thanks for working on this.
> 
> I'm wondering if we should simplify this to remove the cluster deployment resources (or split them into a separate chart). I think it's more likely to deploy this on existing clusters.

@morgante, thanks! Today, there's a chart for load balancing resources and a chart for cluster resources. I think you're suggesting we remove cluster/templates/deployment.yaml from the 2nd chart and have it be applied separately and optionally as example. Is this correct?

morgante commented:
> @morgante, thanks! Today, there's a chart for load balancing resources and a chart for cluster resources. I think you're suggesting we remove cluster/templates/deployment.yaml from the 2nd chart and have it be applied separately and optionally as example. Is this correct?

Right, I'd suggest pulling the cluster creation templates into a separate chart.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+
+All steps are run from this directory.
+
+1. Review and update the values in `./lb/values.yaml`. Note that if you change cluster name and locaiton, you will need to change how they are used in `gcloud container clusters get-credentials` commands below. [PROJECT_ID] refers to the project where all the GCP resources will be created.  
```
nit: "locaiton" -> "location"

AlexBulankou replied:
thanks! fixed

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/cluster/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart to deploy example cluster configuration for GCP multi-cluster ingress
+name: gcp-mci-cluster
```
I found that you've updated the name of this solution but there are still "mci" left in some of the names. Is it intended?

AlexBulankou replied:
not intended. fixed here and in other chart names.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+## REQUIREMENTS
+
+1. GKE Cluster with Config Connector
+1. [Helm](../../../README.md#helm)
```
I'm using KCC installed in project A (a fresh installation of KCC so no other APIs are enabled) to manage GCP resources in project B (a brand new project with nothing configured), and the requirements are not clear to me that:
1. What roles does the "cnrm-system" GSA need to have in project B in order to run the solution?
2. What APIs should I manually enable in project A and/or project B?

AlexBulankou replied:
Thanks. I've added the note that this solution assumes that all resources are installed in the same project, where the cluster with Config Connector is installed, and that load balancing resources are installed on the same cluster where Config Connector is installed. 
It is best to simplify the solution this way, because otherwise multiple steps below become more complicated, also we'll need to test it this way too, which is extra work.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+    helm template ./lb/ --set projectId=[PROJECT_ID]
+
+    # install your chart
+    helm install ./lb/ --set projectId=[PROJECT_ID] --generate-name
```
I got the following error when I run the solution. Is there anything wrong with the role name?

```
Error: IAMPolicyMember.iam.cnrm.cloud.google.com "iampolicymember-autoneg-custom" is invalid: []:
 Invalid value: map[string]interface {}{"apiVersion":"iam.cnrm.cloud.google.com/v1beta1", 
"kind":"IAMPolicyMember", "metadata":map[string]interface {}{"creationTimestamp":"2020-06-
09T18:35:26Z", "generation":1, "name":"iampolicymember-autoneg-custom", "namespace":"
[PROJECT_ID]", "uid":"fc9be478-aa7f-11ea-9c6f-42010a800076"}, "spec":map[string]interface {}
{"member":"serviceAccount:autoneg-system@[PROJECT_ID].iam.gserviceaccount.com", 
"resourceRef":map[string]interface {}{"apiVersion":"resourcemanager.cnrm.cloud.google.com/v1beta1", 
"external":"projects/[PROJECT_ID]", "kind":"Project"}, "role":"projects/[PROJECT_ID]/roles/autonegrole"}}: 
validation failure list:
spec.role in body should match '^roles/[\w\.]+$'
```

AlexBulankou replied:
Thanks. I added an item to requirements.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+
+    # curl the external address of the forwarding rule. Note that it might take around 5-10 minutes for load balancing to start working.
+    # You will see the message ("Hello from North America" or "Hello from Europe" based on your location).
+    curl $(kubectl get  computeforwardingrule -o=jsonpath='{.items[0].spec.ipAddress.addressRef.external}')
```
I got the following error message:
```
$ curl $(kubectl get  computeforwardingrule -o=jsonpath='{.items[0].spec.ipAddress.addressRef.external}') 

<html><head>
<meta http-equiv="content-type" content="text/html;charset=utf-8">
<title>502 Server Error</title>
</head>
<body text=#000000 bgcolor=#ffffff>
<h1>Error: Server Error</h1>
<h2>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.</h2>
<h2></h2>
</body></html>
```

When I only did the kubectl command, it succeeded:
```
$ kubectl get  computeforwardingrule -o=jsonpath='{.items[0].spec.ipAddress.addressRef.external}'
34.120.245.118
```
Did I do anything wrong?

AlexBulankou replied:
I added two steps to readme to check along the way: the first for each of the services in cluster:
```
check the service and ensure that `anthos.cft.dev/autoneg-status` annotation is present in the output
    get svc node-app-backend -o=jsonpath='{.metadata.annotations}'
```
and the second one on the backend service:
```
gcloud compute backend-services describe node-app-backend-service
```

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+    gcloud container clusters get-credentials cluster-na --zone=us-central1-a
+
+    # validate your chart
+    helm lint ./cluster/ --set projectId=[PROJECT_ID] --set localMessage="Hello from North America!"
```
Fixed! (looks like this is non-Mac only btw)

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+    helm lint ./cluster/ --set projectId=[PROJECT_ID] --set localMessage="Hello from North America!"
+
+    # install your chart
+    helm install ./cluster/ --set projectId=[PROJECT_ID] --set localMessage="Hello from North America!" --generate-name
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
````diff
@@ -0,0 +1,142 @@
+
+    ```bash
+    # switch the context to the first cluster. The command uses cluster name and zone based on the values used to create the clusters.
+    gcloud container clusters get-credentials cluster-na --zone=us-central1-a
````
Thanks. Because I added the disclaimer in the requirements - This solution assumes that all resources are installed in the same project, where the cluster with Config Connector is installed, and that load balancing resources are installed on the same cluster where Config Connector is installed, I'm leaving the more complex setup up the consumer to figure out.

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
````diff
@@ -0,0 +1,142 @@
+    kubectl wait --for=condition=Ready pods --all
+    ```
+
+1. Switch the context to the main cluster and run verify that multi-cluster ingress is configured
````
Thanks. Added clarification.

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
````diff
@@ -0,0 +1,142 @@
+1. Switch the context to the main cluster and run verify that multi-cluster ingress is configured
+
+    ```bash
+    # switch the context to the main cluster
````
Thanks. Added note on this.

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+    gcloud container clusters get-credentials [CLUSTER NAME] --zone=[CLUSTER ZONE]
+
+    # curl the external address of the forwarding rule. Note that it might take around 5-10 minutes for load balancing to start working.
+    # You will see the message ("Hello from North America" or "Hello from Europe" based on your location).
```
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,142 @@
+
+    # curl the external address of the forwarding rule. Note that it might take around 5-10 minutes for load balancing to start working.
+    # You will see the message ("Hello from North America" or "Hello from Europe" based on your location).
+    curl $(kubectl get  computeforwardingrule -o=jsonpath='{.items[0].spec.ipAddress.addressRef.external}')
```
fixed.

AlexBulankou commented:
> > @morgante, thanks! Today, there's a chart for load balancing resources and a chart for cluster resources. I think you're suggesting we remove cluster/templates/deployment.yaml from the 2nd chart and have it be applied separately and optionally as example. Is this correct?
> 
> Right, I'd suggest pulling the cluster creation templates into a separate chart.

@morgante , I changed the configuration and added a separate chart to provision the clusters.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,173 @@
+## REQUIREMENTS
+
+1. [Helm](../../../README.md#helm)
+1. GKE Cluster with Config Connector. This solution assumes that all resources are installed in the same project, where the cluster with Config Connector is installed, and that load balancing resources are installed on the same cluster where Config Connector is installed. If you would like to configure your resources in a different project, the easiest approach would be to give your Config Connec ...
```
There is an extra "." after "this": `give your Config Connector service account (`cnrm-system`) owner permissions on this. target project.` Could you remove it?

AlexBulankou replied:
fixed

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,173 @@
+1. [Helm](../../../README.md#helm)
+1. GKE Cluster with Config Connector. This solution assumes that all resources are installed in the same project, where the cluster with Config Connector is installed, and that load balancing resources are installed on the same cluster where Config Connector is installed. If you would like to configure your resources in a different project, the easiest approach would be to give your Config Connec ...
+1. If your Config Connector version is earlier than [1.11.2](https://github.com/GoogleCloudPlatform/k8s-config-connector/releases) you need to apply [this workaround](https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/78#issuecomment-577285402) to `iampolicymembers.iam.cnrm.cloud.google.com` CRD.
+1. `compute.googleapis.com` and `container.googleapis.com` API should be enabled on the project managed by Config Connector, in addition to the default services enabled.
```
API -> APIs?

AlexBulankou replied:
fixed

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,173 @@
+    # if you created the load balancing resources in the namespace, other than default, switch the context to that namespace
+
+    # verify that your backend service has 2 backends attached (select index of "global" if prompted)
+    gcloud compute backend-services describe node-app-backend-service
```
Thank you for adding the additional check, Alex. I tried it, and it looks like my backend services look normal, however I still can't curl the IP address. I checked my load balancing resource in GCP Console, it also looks pretty normal. Maybe we should chat offline to do some debugging together?

AlexBulankou replied:
Sure. I am adding additional information showing how backend service output should look like. Also does this step return correct output for you for both clusters?
```
# check the service and ensure that `anthos.cft.dev/autoneg-status` annotation is present in the output
get svc node-app-backend -o=jsonpath='{.metadata.annotations}'
```
Thanks for offering to help debug! I'll set up a some time to look at it together.

AlexBulankou replied:
@maqiuyujoyce , thanks for helping to debug. I've followed up separately on the details of the error I'm seeing with your setup. I've made changes to the instructions above to enable `cloudresourcemanager.googleapis.com`. 
Can you please try re-running the solution on a new project with this API enabled?

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
````diff
@@ -0,0 +1,190 @@
+      maxRatePerEndpoint: 100.0
+      connectionDraining:
+      drainingTimeoutSec: 300
+    ```
````
Thank you for adding the details! In the output of the command `gcloud compute backend-services describe node-app-backend-service` I saw no backend is attached.

maqiuyujoyce replied:
Had a debugging meeting offline. Now the solution works.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,190 @@
+
+     # switch the context to the cluster that contains the configs for load balancing resources
+    gcloud container clusters get-credentials [CLUSTER NAME] --zone=[CLUSTER ZONE]
+
```
Would be great if you also add the step specified in line 128.

AlexBulankou replied:
Added.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/workload/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart to deploy example workload configuration for GCP autoneg solution
+name: gcp-mci-workload
```
I thought we are going to change the name "mci"?

AlexBulankou replied:
fixed

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/lb/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart to deploy load balancing resources for GCP autoneg solution
+name: gcp-mci-lb
```
I thought we are going to change the name "mci"?

AlexBulankou replied:
fixed.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/workload/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+description: A Helm chart to deploy example workload configuration for GCP autoneg solution
+name: gcp-mci-workload
+version: 0.1.0
+icon: https://upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Globe_icon.svg/1024px-Globe_icon.svg.png
```
A newline would be preferred.

AlexBulankou replied:
fixed

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/clusters/values.yaml:
```diff
@@ -0,0 +1,11 @@
+  name: cluster-eu
+  zone: europe-west2-a
+  masterAuthUser: user
+  masterAuthPwd: password12345678
```
A newline would be preferred.

AlexBulankou replied:
fixed

AlexBulankou commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,190 @@
+    gcloud container clusters get-credentials [CLUSTER NAME] --zone=[CLUSTER ZONE]
+
+    # if you created the load balancing resources in the namespace, other than default, switch the context to that namespace
+
```
Added.

maqiuyujoyce commented on config-connector/solutions/networking/helm/autoneg/README.md:
```diff
@@ -0,0 +1,194 @@
+
+1. [Helm](../../../README.md#helm)
+1. GKE Cluster with Config Connector. This solution assumes that all resources are installed in the same project, where the cluster with Config Connector is installed, and that load balancing resources are installed on the same cluster where Config Connector is installed. If you would like to configure your resources in a different project, the easiest approach would be to give your Config Connec ...
+1. If your Config Connector version is earlier than [1.11.2](https://github.com/GoogleCloudPlatform/k8s-config-connector/releases) you need to apply [this workaround](https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/78#issuecomment-577285402) to `iampolicymembers.iam.cnrm.cloud.google.com` CRD.
```
We don't have version 1.11.2, but instead 1.12.0.

AlexBulankou replied:
fixed.
